### PR TITLE
Try other nodes if command fails

### PR DESCRIPTION
**Pull request by ineiti**
This PR switches to a next node if the current node returns an error. It
tries to keep the same 'working' node as long as possible.

### CODE REVIEW

File: external/js/cothority/src/network/connection.ts [hunk 1]
```diff
@@ -141,25 +149,28 @@ export class RosterWSConnection extends WebSocketConnection {
...
             try {
                 // we need to await here to catch and try another conode
                 return await super.send(message, reply);
             } catch (e) {
-                Log.lvl3(`fail to send on ${addr} with error:`, e);
+                Log.lvl3(`failed to send on ${this.url} with error:`, e);
                 errors.push(e.message);
-                // TODO: send again over all nodes
-                return Promise.reject(e.message);
+                if (i > 2) {
```
Comment (Gilthoniel) [inline external/js/cothority/src/network/connection.ts]: Can we have that as a modifiable parameter of the class ?

File: external/js/cothority/src/network/connection.ts [hunk 2]
```diff
@@ -113,15 +113,23 @@ export class WebSocketConnection implements IConnection {
             });
 
             ws.onClose((code: number, reason: string) => {
-                if (code !== 1000) {
-                    Log.error("Got close:", code, reason);
-                    reject(new Error(reason));
+                switch (code) {
```
[the remark this answers is not in the document]
Comment (ineiti) [reply external/js/cothority/src/network/connection.ts]: the 'normal' close returns error-code `1000`, as in the end of `onMessage`, so I check that, and simply return if it's a 'normal' close.

I changed it now to a simple `if`, as I don't remember why the code `1006` was in there specifically...

File: external/js/cothority/src/network/connection.ts [hunk 3]
```diff
@@ -141,25 +149,28 @@ export class RosterWSConnection extends WebSocketConnection {
...
-                Log.lvl3(`fail to send on ${addr} with error:`, e);
+                Log.lvl3(`failed to send on ${this.url} with error:`, e);
                 errors.push(e.message);
-                // TODO: send again over all nodes
-                return Promise.reject(e.message);
+                if (i > 2) {
+                    return Promise.reject(errors.join(" :: "));
+                }
+                Log.error("Error while sending - trying with next node");
+                this.addresses = this.addresses.slice(1);
```
[the remark this answers is not in the document]
Comment (ineiti) [reply external/js/cothority/src/network/connection.ts]: Even though I have to make a comment about `javascript` and `readable` not going well together ;)

File: external/js/cothority/src/network/connection.ts [hunk 4]
```diff
@@ -113,15 +113,15 @@ export class WebSocketConnection implements IConnection {
             });
 
             ws.onClose((code: number, reason: string) => {
-                if (code !== 1000) {
-                    Log.error("Got close:", code, reason);
-                    reject(new Error(reason));
+                if (code === 1000) {
+                    Log.lvl3("Normal closing of connection");
```
Comment (Gilthoniel) [inline external/js/cothority/src/network/connection.ts]: There's a `return` missing here

Comment (ineiti) [reply external/js/cothority/src/network/connection.ts]: oups - nice catch. Wonder why it passed the tests?

Comment (Gilthoniel) [reply external/js/cothority/src/network/connection.ts]: onClose so I guess the resolve is called before

Comment (ineiti) [reply external/js/cothority/src/network/connection.ts]: so if the `resolve` is called, a later `reject` doesn't work anymore? Makes sense.

**Pull request merged**